[open-] fix opening a dir with an filetype extension

midichef opened the pull request:
If a directory name looks like a file because of its extension (test.txt, test.tsv), Visidata opens it as if it were a file of that type.

`mkdir test.tsv; vd test.tsv`
```
  File "/home/midichef/.venv/lib/python3.12/site-packages/visidata/loaders/tsv.py", line 89, in iterload
    with self.open_text_source() as fp:
         ^^^^^^^^^^^^^^^^^^^^^^^
[...]
  File "/usr/lib/python3.12/pathlib.py", line 1015, in open
    return io.open(self, mode, buffering, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
IsADirectoryError: [Errno 21] Is a directory: 'test.tsv'
```

This PR handles that case, including the case where the filename refers to a symlink to a directory.

midichef committed:
[open-] fix opening a dir with an filetype extension

saulpw reviewed (approved):
Who is even doing this? :)  Thanks for fixing it!

anjakefala commented on visidata/_open.py:
```diff
@@ -108,6 +108,9 @@ def openPath(vd, p, filetype=None, create=False):
         return newfunc(p)
 
+    if p.is_dir():  # also matches symlinks to dirs
+        return vd.open_dir(p)
```
This is breaking the vgit tests! They operate on a dir but load them with `--filetype` git. Is there any way to accomodate that usecase?

midichef replied:
Oops! Okay, I've modified the PR to ignore the extension only for directories, and to accept the filetype specified by the user.

frosencrantz commented:
There is also an issue with the filetype guessers and a path that is a directory.  The guessers should not report errors if the path is a directory.

midichef commented:
@saulpw 
Can you take a careful look at my changes? I don't understand most option handling in visidata, and I don't want to break anything before the 3.1 release.

I've moved two `filetype` assignments into `openPath()` so the logic is all in one place. And I changed the fallback from `''` to `None` to match the default parameter of `openPath()`.

What is `'global'` in `options.getonly`?
https://github.com/saulpw/visidata/blob/d50a1bf9e0f1a65c8cf65165fc2df7dbe7e4c2b3/visidata/_open.py#L151
And how does it differ from vd.options.filetype:
https://github.com/saulpw/visidata/blob/d50a1bf9e0f1a65c8cf65165fc2df7dbe7e4c2b3/visidata/_open.py#L98

Side note, I didn't know we could assign a filetype to a file by its path, like `options.set('filetype', 'csv', '-')`, that's very cool!

midichef commented:
And what should `guess_*` return to indicate it has no guess for a file: `{}` or `None`? I did `{}` for now.

midichef commented:
Oh, I see the existing behavior of `guess_*` in the case of files that did not match the guess, was to return `None`. I've pushed a change to do that instead of return `{}`.

saulpw commented:
> What is 'global' in options.getonly?

This is if the option has been set, not as a default or on a specific sheet, but globally by the user (visidatarc or cli or from options sheet).  Whereas options.filetype normally looks at a hierarchy of places for the option to be set, either on the specific sheet, or the sheet class, or by the user, and finally falling back to the default.

saulpw commented on visidata/loaders/archive.py:
```diff
@@ -10,11 +10,13 @@
 
 @VisiData.api
 def guess_zip(vd, p):
+    if p.is_dir(): return
```
This shouldn't be required for guessers; we let it raise the exception and it should be silently ignored.

midichef replied:
Okay. My aim was to clean up the `--debug` results. But very few people will ever see those. And I am seeing that visidata style is to take full advantage of ignoring exceptions in such situations, and keep the code lean.

saulpw commented on visidata/_open.py:
```diff
@@ -81,6 +81,10 @@ def guess_extension(vd, path):
     If True, *create* will return a new, blank **Sheet** if file does not exist.'''
+    # allow user to assign a filetype to a pathname:  options.set('filetype', 'csv', '-')
+    filetype = filetype or vd.options.getonly('filetype', str(p), None)  #1710
+    filetype = filetype or vd.options.getonly('filetype', 'global', None)
```
Why were these lines moved?  It seems like code at line 102/103 should fix the issue by itself.

midichef replied:
The code at line 102 would fix it by itself. Lines 85, 86, and 104 are all similar in filling in filetype based on options. The intention was to put them together in one function, to make it easier to trace the logic that sets filetype based on options.

saulpw reviewed (changes requested):
The core fix looks reasonable, can we minimize the diff?

anjakefala commented:
@midichef Is your expectation for this PR that now `vd test.tsv -f dir` will work, but `vd test.tsv` might still have a `IsADirectoryError: [Errno 21] Is a directory: 'test.tsv'`?

midichef commented:
> @midichef Is your expectation for this PR that now `vd test.tsv -f dir` will work, but `vd test.tsv` might still have a `IsADirectoryError: [Errno 21] Is a directory: 'test.tsv'`?

Oops, no. I've force-pushed a fix with this diff as the only change:
```
 def guess_dir(vd, p):
     if p.is_dir():
-        return dict(filetype='dir')
+        return dict(filetype='dir', _likelihood=10)
```
In the case of `vd test.tsv`, there were conflicting guesses for file type. `guess_tsv()` was saying `tsv` with likelihood `3`, and `guess_dir()` was saying `dir` with no likelihood, which is interpreted as `1`. Instead, the new version of `guess_dir()` uses a high priority to override. `10` matches the highest priority currently used in visidata, seen in a few loaders:  `api_airtable.py`, `http.py` and `xlsb.py`.

anjakefala commented:
Thanks @midichef! =)
